internal/container/environment.go

```
@@ -0,0 +1,116 @@
+	tmpFile string
+	// The client used to fetch secrets from Secret Manager, if any.
+	secretManagerClient *secretmanager.Client
+	// The project in which to look up secrets
+	secretsProject string
+	// A cache of secrets we've already fetched.
+	secretCache map[string]string
+	// The pipeline configuration, specifying which environment variables to obtain
+	// for each command.
+	pipelineConfig *statepb.PipelineConfig
+}
+
+func NewEnvironment(ctx context.Context, tmpRoot, secretsProject string, pipelineConfig *statepb.PipelineConfig) (*ContainerEnvironment, error) {
+	var secretManagerClient *secretmanager.Client
+	if secretsProject != "" {
+		client, err := secretmanager.NewClient(ctx)
+		if err != nil {
+			return nil, err
+		}
+		secretManagerClient = client
+	} else {
+		secretManagerClient = nil
+	}
+	tmpFile := filepath.Join(tmpRoot, "docker-env.txt")
+	return &ContainerEnvironment{
+		ctx:                 ctx,
+		tmpFile:             tmpFile,
+		secretManagerClient: secretManagerClient,
+		secretsProject:      secretsProject,
+		secretCache:         make(map[string]string),
+		pipelineConfig:      pipelineConfig,
+	}, nil
+}
+
+func writeEnvironmentFile(containerEnv *ContainerEnvironment, commandName string) error {
+	content, err := constructEnvironmentFileContent(containerEnv, commandName)
+	if err != nil {
+		return err
+	}
+	return utils.CreateAndWriteToFile(containerEnv.tmpFile, content)
+}
+
+func constructEnvironmentFileContent(containerEnv *ContainerEnvironment, commandName string) (string, error) {
+	commandConfig := containerEnv.pipelineConfig.Commands[commandName]
+	if commandConfig == nil {
+		return "# No environment variables", nil
+	}
+	var builder strings.Builder
+	for _, variable := range commandConfig.EnvironmentVariables {
+		value, present := os.LookupEnv(variable.Name)
+		if present {
+			builder.WriteString(fmt.Sprintf("%s=%s\n", variable.Name, value))
+			continue
+		}
+		if containerEnv.secretManagerClient == nil {
```

this is where having a type might be helpful, if it is type secret and this is null should we return an error.

---

The choice at the moment leaves it to the container to decide whether the value is actually required or not. We could add required/optional in the config file, but I think it's probably simpler to leave it up to the container (especially as it may only be required for some APIs).